cpp/src/arrow/compute/kernels/mask.cc

```diff
@@ -0,0 +1,230 @@
+          typename OutBuilder>
+Status MaskImpl(FunctionContext*, const ValueArray& values, const BooleanArray& mask,
+                OutBuilder* builder) {
+  auto offset = mask.offset();
+  auto length = mask.length();
+  internal::BitmapReader mask_data(mask.data()->buffers[1]->data(), offset, length);
+  for (int64_t i = 0; i < mask.length(); mask_data.Next(), ++i) {
+    if (!WholeMaskValid && mask.IsNull(i)) {
+      builder->UnsafeAppendNull();
+      continue;
+    }
+    if (mask_data.IsNotSet()) {
+      continue;
+    }
+    if (!AllValuesValid && values.IsNull(i)) {
+      builder->UnsafeAppendNull();
+      continue;
+    }
+    RETURN_NOT_OK(UnsafeAppend(builder, values.GetView(i)));
+  }
+  return Status::OK();
+}
+
+template <bool AllValuesValid, typename ValueArray, typename MaskArray,
+          typename OutBuilder>
+Status UnpackMaskNullCount(FunctionContext* context, const ValueArray& values,
+                           const MaskArray& mask, OutBuilder* builder) {
+  if (mask.null_count() == 0) {
+    return MaskImpl<AllValuesValid, true>(context, values, mask, builder);
+  }
+  return MaskImpl<AllValuesValid, false>(context, values, mask, builder);
+}
+
+template <typename ValueArray, typename MaskArray, typename OutBuilder>
+Status UnpackValuesNullCount(FunctionContext* context, const ValueArray& values,
+                             const MaskArray& mask, OutBuilder* builder) {
+  if (values.null_count() == 0) {
+    return UnpackMaskNullCount<true>(context, values, mask, builder);
+  }
+  return UnpackMaskNullCount<false>(context, values, mask, builder);
+}
+
+template <typename T>
+using ArrayType = typename TypeTraits<T>::ArrayType;
+
+template <typename MaskType>
+struct UnpackValues {
+  template <typename ValueType>
+  Status Visit(const ValueType&) {
+    using OutBuilder = typename TypeTraits<ValueType>::BuilderType;
+    auto&& mask = static_cast<const ArrayType<MaskType>&>(*params_.mask);
+    auto&& values = static_cast<const ArrayType<ValueType>&>(*params_.values);
+    std::unique_ptr<ArrayBuilder> builder;
+    RETURN_NOT_OK(MakeBuilder(params_.context->memory_pool(), values.type(), &builder));
+    RETURN_NOT_OK(builder->Reserve(OutputSize(mask)));
+    RETURN_NOT_OK(UnpackValuesNullCount(params_.context, values, mask,
+                                        static_cast<OutBuilder*>(builder.get())));
+    return builder->Finish(params_.out);
+  }
+
+  Status Visit(const NullType& t) {
+    auto&& mask = static_cast<const ArrayType<MaskType>&>(*params_.mask);
+    params_.out->reset(new NullArray(OutputSize(mask)));
+    return Status::OK();
+  }
+
+  Status Visit(const DictionaryType& t) {
+    std::shared_ptr<Array> masked_indices;
+    const auto& values = internal::checked_cast<const DictionaryArray&>(*params_.values);
+    {
+      // To take from a dictionary, apply the current kernel to the dictionary's
+      // mask. (Use UnpackValues<MaskType> since MaskType is already unpacked)
```

Comment:
You mean "to the dictionary's indices" I suppose?

Reply:
I do, thanks